### ISSUE TITLE
Respect specified versions in an update all

### ISSUE THREAD

**Issue by dhirschfeld** [conda/conda#4394]
With conda 4.2.15 you can specify the versions for specific packages in an `update --all`
```
C:\dev\code\sandbox> conda list ^conda$
# packages in environment at C:\bin\Anaconda:
#
conda                     4.2.15                   py35_0
C:\dev\code\sandbox> conda update icu=57.1 qt=5.6 conda=4.2.15 --all
Fetching package metadata ...............
Solving package specifications: ..........

# All requested packages already installed.
```
With conda 4.3.6 the version specifications are ignored:
```
C:\dev\code\sandbox> conda update icu=57.1 qt=5.6 --all
Fetching package metadata .................
Solving package specifications: .

Package plan for installation in environment C:\bin\Anaconda:

The following packages will be UPDATED:

    icu:                57.1-vc14_0                     [vc14] --> 58.1-vc14_0       conda-forge [vc14]
    matplotlib:         2.0.0b4-np111py35_0 conda-forge --> 2.0.0-np111py35_0 conda-forge

The following packages will be DOWNGRADED due to dependency conflicts:

    anaconda-navigator: 1.3.2-py35_0                    --> 1.2.3-py35_0
    libxml2:            2.9.4-vc14_2        conda-forge [vc14] --> 2.9.3-vc14_9      conda-forge [vc14]
    pyqt:               5.6.0-py35_2                    --> 4.11.4-py35_7
    qt:                 5.6.2-vc14_3                    [vc14] --> 4.8.7-vc14_6      conda-forge [vc14]
    qtconsole:          4.2.1-py35_2                    --> 4.2.1-py35_0

Proceed ([y]/n)? n
```

**Comment by kalefranz** [comment]
To my knowledge, update has *never* respected versions specs. Or more accurately, I had thought version specs were invalid in an update operation. In the example you're giving, we're complicating that understanding with the --all flag, which has consistently proven to mean different things to different people. Or alternately, different people have different intuitive expectations of what it should do. 

All that said, I don't have a good explanation for what you're seeing. I'm going to make a test case out of it. Whether the result we land on is the one you expected, we'll at least have a consistent outcome from this point forward.

**Comment by kalefranz** [comment]
CC @mcg1969

**Comment by mcg1969** [comment]
You're not supposed to be able to specify versions at all in a `conda update` command.

**Comment by mcg1969** [comment]
You seem to have accidentally stumbled upon a feature that was never supposed to be there. And I am afraid that [I am the cause](https://github.com/conda/conda/commit/123849f484ddb3eb86fc7409a980b1346aa398fd).

In [this line](https://github.com/conda/conda/blob/master/conda/cli/install.py#L140), you see that `conda` is instructed to reject update specs that are anything but bare package names (well, you'd have to look at the definition of `arg2spec` to see for sure). 

So what was supposed to happen when `--all` is present? Well, in [conda 3.x](https://github.com/conda/conda/blob/3.x/conda/cli/install.py#L132), it was illegal to supply packages alongside `--all`. This is about the time I began working on conda.

In that first commit above, I revamp the `conda update` functionality, and I inadvertently [removed the check](https://github.com/conda/conda/commit/123849f484ddb3eb86fc7409a980b1346aa398fd#diff-34bb8d01f991a5ffabae5e80b987c83fL130) that ensured that package specs and`--all` do not mix. I can assure you that it was inadvertent.

Honestly, I sort of like the behavior this causes for `--all`, but it makes no sense for it to be there for `--all` and not without it. So we either allow it for all update commands or we remove it here.

**Comment by dhirschfeld** [comment]
I actually really like being able to fix specific versions in an update all - in this case it allows me to pin icu and qt since the updated icu will cause my qt/matplotlib to segfault and I don't want to downgrade qt.

I supposed I could *actually* pin those packages but its' convenient to be able to do it at the command line rather than by creating a file somewhere I can never remember...

I guess it's fine if it's not supported in which case it should probably error out with an explanation- I just noticed the discrepancy so thought I bring it up.

**Comment by mingwandroid** [comment]
Is there a bug filed for the matplotlib segfault?

**Comment by mcg1969** [comment]
@dhirshfeld  I have to admit, I like the behavior, and I see its usefulness.  I would actually be willing to allow it to come back for any update command, even without --all.

**Comment by mcg1969** [comment]
But I do defer to @kalefranz on this.

**Comment by kalefranz** [comment]
I'm not opposed to `conda update icu=57.1 qt=5.6 --all`.  I kind of like it too.  I've got a WIP PR targeted for 4.4.x started where I want to clean up `conda/cli/install.py` and `conda/plan.py` (which became even more unwieldy in 4.3.x).  I can try to work this in as part of that.

**Comment by dhirschfeld** [comment]
@mingwandroid - it's not a bug per-se; qt 5.6.2 was built against icu 57.* and won't load the dll without that specific version. Although qt does specify 57.* in the requirements, somehow I'd ended up with the later version which cause a few headaches tracking down. Probably some combination of force installing/un-installing mucked up my deps though I'll note that conda doesn't seem to have a problem upgrading to the latest version if asked to even though I have qt 5.6.2 in my env:
```
C:\dev\code\sandbox> conda list ^qt$
# packages in environment at C:\bin\Anaconda:
#
qt                        5.6.2                    vc14_3  [vc14]
C:\dev\code\sandbox> conda info qt=5.6.2=vc14_3
Fetching package metadata ...............

qt 5.6.2 vc14_3
---------------
file name   : qt-5.6.2-vc14_3.tar.bz2
name        : qt
version     : 5.6.2
build number: 3
build string: vc14_3
channel     : defaults
size        : 55.3 MB
date        : 2017-01-16
features    : vc14
fn          : qt-5.6.2-vc14_3.tar.bz2
license     : LGPLv3
license_family: LGPL
md5         : 89faebaf05110dfab65383ed09e14376
priority    : 0
schannel    : defaults
url         : https://repo.continuum.io/pkgs/free/win-64/qt-5.6.2-vc14_3.tar.bz2
dependencies:
    icu 57.*
    jpeg 9*
    libpng >=1.6.27,<1.7
    openssl 1.0*
    zlib 1.2*
C:\dev\code\sandbox> conda update icu
Fetching package metadata ...............
Solving package specifications: ..........

Package plan for installation in environment C:\bin\Anaconda:

The following packages will be downloaded:

    package                    |            build
    ---------------------------|-----------------
    conda-4.3.7                |           py35_0         533 KB

The following packages will be UPDATED:

    conda: 4.2.15-py35_0 --> 4.3.7-py35_0
    icu:   57.1-vc14_0   [vc14] --> 58.1-vc14_0  conda-forge [vc14]

Proceed ([y]/n)? n

Exiting
```

**Comment by mingwandroid** [comment]
It looks like a bug to me. The solver should not update ICU like this as it will break qt. If you can rerun with -v -v -v a paste the output that's be good.

As an aside, missing conda-forge and defaults in the same env is something is would strongly discourage as things stand.

@mcg1969, your thoughts?

**Comment by mingwandroid** [comment]
It looks like a bug to me. The solver should not update ICU like this as it will break qt. If you can rerun with -v -v -v a paste the output that's be good.

As an aside, mixing conda-forge and defaults in the same env is something is would strongly discourage as things stand.

@mcg1969, your thoughts?

**Comment by mcg1969** [comment]
The solver is never wrong. i'm not kidding; it has proven itself to us in countless situations like these. The metadata is the problem. 

In a case like this it is likely one of two things. One, due to differences in metadata choices between conda-forge and defaults. 

Or two, due to an environment that has existing conflicts in it. In that case, the `bad_installed` functionality in `resolve.py` is forced to limit the solver only to the immediate set of packages that are involved in a given conda command, instead of ensuring that the entire environment becomes consistent. This is a throwback to conda 3.x, which did _not_ properly preserve environment consistency, and the need for conda to work gracefully in such broken environments. I do not like this functionality, and as I recently said in another thread, I think it either needs to be dramatically improved or removed.

To see whether it is the second problem, we need to see the full `--debug` output of the conda install command. Look for a message like (paraphrasing) "Limiting the solver to the following packages".

**Comment by mcg1969** [comment]
You can disregard my first scenario above; I now see that icu is being updated, and it's Qt that is supposed to be limiting it. That leaves the second scenario.

**Comment by dhirschfeld** [comment]
It's likely my env is in an inconsistent state - I'll try posting the debug output when I'm back on that machine.

Whilst it's nice in theory to always have a perfectly consistent environment in some cases the metadata for a package is just *wrong* and in those cases you do need to be able to tell conda to ignore the deps and just bloody well install the package! 

NB: `conda install --force` currently works well for this use-case but may lead to other undesirable behaviour such as above. Maybe it would be good to have a flag `--no-deps` which installed just the package but *also* marked it in such a way that it's dependencies would never be considered for all future conda commands.

The specific package giving me grief is matplotlib which has a hard dependency on qt4 whereas I want to run against qt5 my workaround is to install matplotlib with pip so the dependencies aren't managed by conda - I just have to make sure conda doesn't upgrade my icu!

This particular problem is I believe a large reason why *optional dependencies* have been requested (#793, #1696, #3299)

**Comment by kalefranz** [comment]
Conditional dependencies will be in 4.4!

**Comment by mcg1969** [comment]
Yep, we already have the code to do optional dependencies.

> Whilst it's nice in theory to always have a perfectly consistent environment in some cases the metadata for a package is just wrong and in those cases you do need to be able to tell conda to ignore the deps and just bloody well install the package!

This is fine, but the problem is that conda cannot _possibly_ be clairvoyant enough to know which inconsistencies it is supposed to ignore and which it is supposed to fix. So it necessarily has to limit the scope of its solver, and that means sometimes it's going to make things worse.

The only thing I can think to do here is to declare one or more packages as _pinned and ignorable_ by conda. That way, the user can explicitly state that an inconsistency in a specific package can be ignored. We could go a bit further and say: if the package _can_ be made consistent, do it; but if not, leave it alone. (Thanks to optional dependencies for that one!)

**Comment by mcg1969** [comment]
I just attempted to improve `bad_installed` by including within its scope the largest subset of consistent installed packages that `conda` can find. Hopefully that prevents `conda` from making a broken environment even _more_ broken, while at the same time providing reasonable results in those same broken environments.

**Comment by dhirschfeld** [comment]
In case it's of use the debug output from the update of icu (which shouldn't be possible) is available at:
https://gist.github.com/dhirschfeld/1555c8c2b648d6d69a19bb7cde2418bf

cc: @mingwandroid, @mcg1969

**Comment by mcg1969** [comment]
Yep, search for `bad_installed` in the log. The environment is already inconsistent so it is reverting to effectively conda 3.x behavior.

**Comment by dhirschfeld** [comment]
I propose a new command - `conda env fix` :laughing: 

Seriously though, how can I fix my environment? How can I even know which packages are in conflict?

I suppose I can always just nuke the whole install and start from scratch with conda 4.3.8+ but it might be nice if there were a less drastic way to ensure consistency...

**Comment by kalefranz** [comment]
I have plans for something exactly like that :)

**Comment by mcg1969** [comment]
@dhirschfeld we've merged #4402 which attempts to be more aggressive about the scope of packages it considers in inconsistent environments. I think this should have a positive impact in this situation. Care to try it? I think it might even help environments gracefully heal themselves.

**Comment by dhirschfeld** [comment]
With 4.3.8 it no longer tries to update icu and generally seems to be working well for me so far :)
